spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -240,48 +277,90 @@ case class DeltaSharingSource(
       isStartingVersion: Boolean,
       currentLatestVersion: Long): Unit = {
     lastQueryTableTimestamp = System.currentTimeMillis()
+    minUrlExpirationTimestamp = -1
     if (isStartingVersion) {
       // If isStartingVersion is true, it means to fetch the snapshot at the fromVersion, which may
       // include table changes from previous versions.
       val tableFiles = deltaLog.client.getFiles(
         deltaLog.table, Nil, None, Some(fromVersion), None, None
       )
       latestRefreshFunc = () => {
-        deltaLog.client.getFiles(
+        val files = deltaLog.client.getFiles(
           deltaLog.table, Nil, None, Some(fromVersion), None, None
-        ).files.map { f =>
+        ).files
+        var minUrlExpiration: Long = -1L
+        val idToUrl = files.map { f =>
+          if (f.expirationTimestamp != null) {
+            if (minUrlExpiration == -1L) {
+              minUrlExpiration = f.expirationTimestamp
+            } else {
+              minUrlExpiration.min(f.expirationTimestamp)
+            }
+          }
           f.id -> f.url
         }.toMap
+
+        refreshSortedFetchedFiles(idToUrl)
+
+        (idToUrl, minUrlExpiration)
       }
 
       val numFiles = tableFiles.files.size
       tableFiles.files.sortWith(fileActionCompareFunc).zipWithIndex.foreach {
-        case (file, index) if (index > fromIndex) =>
-          appendToSortedFetchedFiles(
-            IndexedFile(
-              fromVersion,
-              index,
-              AddFileForCDF(
-                file.url,
-                file.id,
-                file.partitionValues,
-                file.size,
+        case (file, index) =>
+          if (file.expirationTimestamp != null) {
+            if (minUrlExpirationTimestamp == -1) {
+              minUrlExpirationTimestamp = file.expirationTimestamp
+            } else {
+              minUrlExpirationTimestamp.min(file.expirationTimestamp)
+            }
+          }
+          if (index > fromIndex) {
+            appendToSortedFetchedFiles(
+              IndexedFile(
                 fromVersion,
-                file.timestamp,
-                file.stats
-              ),
-              isLast = (index + 1 == numFiles)))
+                index,
+                AddFileForCDF(
+                  file.url,
+                  file.id,
+                  file.partitionValues,
+                  file.size,
+                  fromVersion,
+                  file.timestamp,
+                  file.stats,
+                  file.expirationTimestamp
+                ),
+                isLast = (index + 1 == numFiles))
+            )
+          }
         // For files with index <= fromIndex, skip them, otherwise an exception will be thrown.
         case _ => ()
       }
+      if (!CachedTableManager.INSTANCE.isValidUrlExpirationTime(minUrlExpirationTimestamp)) {
+        // reset to -1 to indicate that it's not a valid url expiration timestamp.
```

Comment:
Is this correct?
Since this is returned by the server, we should simply trust it and do refresh as needed (perhaps log an error).
Otherwise the urls will expire anyways.

Reply:
hmmm, if this is set to a timestamp expiring very soon and we always respect it, then the client will be refreshing endlessly, refresh after refresh...
Is that desired?